clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -338,6 +384,197 @@ impl FunctionType {
+            TypeSignature::TupleType(atom_tuple_type),
+            TypeSignature::TupleType(expected_tuple_type),
+        ) => {
+            for (name, atom_field_type) in atom_tuple_type.get_type_map() {
+                match expected_tuple_type.field_type(name) {
+                    Some(expected_field_type) => {
+                        inner_type_check_type(
+                            db,
+                            contract_context,
+                            atom_field_type,
+                            expected_field_type,
+                            depth + 1,
+                        )?;
+                    }
+                    None => {
+                        return Err(CheckErrors::TypeError(
+                            expected_type.clone(),
+                            atom_field_type.clone(),
+                        )
+                        .into())
+                    }
+                }
+            }
+        }
+        (
+            TypeSignature::TraitReferenceType(atom_trait_id),
+            TypeSignature::TraitReferenceType(expected_trait_id),
+        ) => {
+            if atom_trait_id != expected_trait_id {
+                let atom_trait = lookup_trait(db, contract_context, &atom_trait_id)?;
+                let expected_trait = lookup_trait(db, contract_context, expected_trait_id)?;
+                trait_check_trait_compliance(
+                    db,
+                    contract_context,
+                    &atom_trait_id,
+                    &atom_trait,
+                    expected_trait_id,
+                    &expected_trait,
+                )?;
+            }
+        }
+        (TypeSignature::NoType, _) => (),
+        (_, _) => {
+            if !expected_type.admits_type(&actual_type) {
+                return Err(
+                    CheckErrors::TypeError(expected_type.clone(), actual_type.clone()).into(),
+                );
+            }
+        }
+    }
+    Ok(expected_type.clone())
+}
+
+fn lookup_trait(
+    db: &mut AnalysisDatabase,
+    contract_context: Option<&ContractContext>,
+    trait_id: &TraitIdentifier,
+) -> CheckResult<BTreeMap<ClarityName, FunctionSignature>> {
+    // If the trait is from the current contract, its not in the db yet.
+    if let Some(contract_context) = contract_context {
+        if contract_context.is_contract(&trait_id.contract_identifier) {
+            return Ok(contract_context
+                .get_trait(&trait_id.name)
+                .ok_or(CheckErrors::NoSuchTrait(
+                    trait_id.contract_identifier.to_string(),
+                    trait_id.name.to_string(),
+                ))?
+                .clone());
+        }
+    }
+
+    db.get_defined_trait(&trait_id.contract_identifier, &trait_id.name)?
+        .ok_or(
+            CheckErrors::NoSuchTrait(
+                trait_id.contract_identifier.to_string(),
+                trait_id.name.to_string(),
+            )
+            .into(),
+        )
```

Comment:
Looking at this, I wonder if it may be a better idea to store the "included" trait in the `contract_context` when handling `use-trait`, that way, you won't need to use the DB to perform trait resolution. The analysis will need to add additional memory usage to the cost_tracker, but it would cut down on the number of DB operations needed to validate trait usage.

Reply:
That suggestion makes sense to me.

Reply:
Yeah, when we see full blocks, they're almost always full due to MARF reads.  We should definitely strive to cut down on them.

Reply:
These traits are saved in the contract context and retrieved with the `lookup_trait` method.